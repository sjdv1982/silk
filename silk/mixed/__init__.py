"""
Python types inside Numpy arrays must be containers, strings, or None, not scalars.
Typedefs are either a string or a typedict with a "type" field containing such a string.
Type strings can be JSON schema types or "tuple", in which case "shape" must be defined
An "array" or "tuple" type contains a field "items", and a field "identical". If "identical" is True, then "items" contains only a single type/typedict,
"object" typedef always contain a field "properties"
A NumPy array (or tuple) of Python objects also has a field "item_storage", which can be "pure-plain" or "mixed-plain", depending on the Python type
If "identical" is False, "item_storage" is a list
In a typedict, a child's "storage" is stored by the parent into a child typedict, but only if:
   1) the child is not pure; or 2) the child is binary and the parent is plain (or vice versa)
"""

MAGIC_NUMPY = b"\x93NUMPY"
MAGIC_SEAMLESS = b'\x93SEAMLESS'
MAGIC_SEAMLESS_MIXED = b'\x94SEAMLESS-MIXED'

import numpy as np
from functools import partialmethod

from ..SilkBase import binary_special_method_names
from .. import Wrapper
from ..validation import (
  _array_types, _integer_types, _float_types, _string_types, _unsigned_types,
  _allowed_types, Scalar, is_np_struct, FormWrapper
)

scalars = ("boolean", "integer", "number", "string")

import numpy as np

def is_numpy_buffer(buffer):
    if buffer is None:
        return False
    return buffer[:len(MAGIC_NUMPY)] == MAGIC_NUMPY

class MonitorTypeError(TypeError):
    pass

class MixedBase(Wrapper):
    def __init__(self, _monitor, _path):
        self._monitor = _monitor
        self._path = _path
    def _unwrap(self):
        return FormWrapper(self.value, self.form, self.storage)
    @property
    def value(self):
        data = self._monitor.get_data(self._path)
        return data
    @property
    def form(self):
        return self._monitor.get_form(self._path)
    @property
    def storage(self):
        return self._monitor.get_storage(self._path)
    def set(self, value):
        self._monitor.set_path(self._path, value)
    def __setattr__(self, attr, value):
        if attr.startswith("_"):
            return super().__setattr__(attr, value)
        return self.__setitem__(attr, value)
        raise AttributeError(attr)
    def __getattr__(self, attr):
        if attr.startswith("_"):
            raise AttributeError(attr)
        if isinstance(self.value, np.ndarray):
            return getattr(self.value, attr)
        else:
            return self.__getitem__(attr)
        raise AttributeError(attr)
    def __str__(self):
        return str(self.value)
    def __repr__(self):
        return repr(self.value)

def mixed_scalar_binary_method(self, other, name):
    if isinstance(other, MixedBase):
        other = other.value
    return getattr(self.value,name)(other)

def mixed_scalar_binary_method_inplace(self, other, name2):
    result = getattr(self.value,name2)(other)
    if result is NotImplemented:
        return NotImplemented
    self._monitor.set_path(self._path, result)
    return result

class MixedScalar(MixedBase):
    def __getitem__(self, item):
        value = self.value
        if not isinstance(value, (str, np.ndarray)):
            raise TypeError(type(value))
        return value[item]
    def __getattr__(self, attr):
        if attr.startswith("_"):
            raise AttributeError(attr)
        if isinstance(self.value, np.ndarray):
            return getattr(self.value, attr)
        elif isinstance(attr, str):
            try:
                return self.__getitem__(attr)
            except TypeError:
                raise AttributeError(attr) from None
        raise AttributeError(attr)


for name in binary_special_method_names:
    if name.startswith("__i"):
        name2 = "__" + name[3:]
        m = partialmethod(mixed_scalar_binary_method_inplace, name2=name2)
    else:
        m = partialmethod(mixed_scalar_binary_method, name=name)
    setattr(MixedScalar, name, m)


from .MixedDict import MixedDict
from .MixedList import MixedList
from .MixedObject import MixedObject
from .Monitor import Monitor
from .Backend import Backend, DefaultBackend, SilkBackend
from .get_form import is_contiguous, is_unsigned
