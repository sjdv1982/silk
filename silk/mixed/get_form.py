"""
Paradigm:
- get_form_XXX as long as the data is in plain storage mode
- When we encounter a numpy array Q, we go into binary storage mode:
    - First determine the form based on dtype: get_tform_XXX
    - Skip Python objects inside the dtype (storage = None)
    - Then, visit (= fill in) those Python objects using the data: visit_typedef_XXX
    - Re-enter plain mode as soon as we encounter these Python objects: get_form_XXX
    - After each Python object, leave plain-mode and return to binary
    - After finishing the Numpy array Q, leave binary mode and return to plain
"""
import numpy as np
from numpy import ndarray, void
from copy import deepcopy
from . import ( Scalar,
  _array_types, _integer_types, _float_types, _string_types, _unsigned_types
)
_string_types = (str,)  ##JSON cannot deal with bytes

dt_builtins = (
    np.dtype("bool"),
    np.dtype("int8"),
    np.dtype("int16"),
    np.dtype("int32"),
    np.dtype("int64"),
    np.dtype("uint8"),
    np.dtype("uint16"),
    np.dtype("uint32"),
    np.dtype("uint64"),
    np.dtype("float32"),
    np.dtype("float64"),
)
def is_contiguous(data):
    """Returns if the data is C-contiguous
    The last stride must be itemsize (bytesize),
     the second-last must be itemsize * shape[-1], etc.
    These criteria are stricter than Numpy's c_contiguous:
      see the documentation of numpy.ndarray.flags for an explanation
    """
    if not data.flags["C_CONTIGUOUS"]:
        return False
    stride = data.itemsize
    contiguous = False
    for n in range(data.ndim-1, -1, -1):
        if data.strides[n] != stride:
            break
        stride *= data.shape[n]
    else:
        contiguous = True
    return contiguous

def is_unsigned(dt):
    return any([dt == t for t in _unsigned_types])

def get_typedef_scalar(value):
    if isinstance(value, bool):
        typedef = "boolean"
    elif isinstance(value, _integer_types):
        typedef = "integer"
    elif isinstance(value, _float_types):
        typedef = "number"
    elif isinstance(value, _string_types):
        typedef = "string"
    elif value is None:
        typedef = "null"
    else:
        raise TypeError(type(value))
    return typedef


def visit_typedef_numpy_items(item_typedef, data):
    #numpy items of mixed-binary storage
    identical = True
    items = item_typedef
    items2 = []
    for d in data:
        item_typedef_curr = deepcopy(item_typedef)
        _ = visit_typedef_numpy("mixed-binary", item_typedef_curr, d)
        items2.append(item_typedef_curr)
    items = items2[0]
    for item_typedef_curr in items2[1:]:
        if item_typedef_curr != items2[0]:
            identical = False
            items = items2
            break
    return items, identical

def visit_typedef_numpy_tuple(storage0, typedef, data):
    # An Numpy tuple field of mixed-binary
    assert storage0 == "mixed-binary" #If None, call visit_typedef_numpy instead
    assert typedef["identical"]
    assert typedef["type"] == "tuple"

    assert data.shape == typedef["shape"]
    item_typedef = typedef["items"]
    items, identical = visit_typedef_numpy_items(item_typedef, data)
    typedef["identical"] = identical
    typedef["item_storage"] = item_storage
    typedef["items"] = item_typedef
    return "mixed-binary"

def visit_typedef_numpy_struct(storage0, typedef, data):
    assert "properties" in typedef ##since type is "object"
    props = typedef["properties"]
    storage = "pure-binary"
    for k in props:
        subtypedef = props[k]
        if not "storage" in subtypedef:
            continue
        subdata = data[k]
        substorage0 = subtypedef["storage"]
        substorage = visit_typedef_numpy(substorage0, subtypedef, subdata)
        if substorage0 is None or substorage != "pure-binary":
            storage = "mixed-binary"
        subtypedef["storage"] = substorage
    if storage == "mixed-binary":
        for k in props:
            subtypedef = props[k]
            substorage = subtypedef.get("storage")
            if substorage is None:
                subtypedef["storage"] = "pure-binary"
            elif substorage == "mixed-binary":
                subtypedef.pop("storage")
    return storage

def visit_typedef_numpy(storage0, typedef, data):
    #visit and fill in mixed-binary typedefs
    if storage0 is None: #typedef of a Python object inside Numpy
        assert not isinstance(data, void), type(data)
        if isinstance(typedef, dict) and "items" in typedef:
            assert typedef["items"] is None
            storage = "mixed-binary"
            if typedef["type"] in ("array", "tuple"): # Numpy array of Python objects
                if typedef["type"] == "tuple": # Numpy multi-element field of Python objects
                    assert data.shape == typedef["shape"]
                item_storage, item_typedef = [], []
                for d in data: #TODO: multi-dimensional (currently NotImplementedError)
                    curr_item_storage, curr_item_typedef = get_form_python_inside_numpy(d)
                    item_storage.append(curr_item_storage)
                    item_typedef.append(curr_item_typedef)
                for dnr, d in enumerate(data[1:]):
                    if item_storage[dnr] != item_storage[0] or \
                      item_typedef[dnr] != item_typedef[0]:
                        identical = False
                        break
                else:
                    identical = True
                    item_storage = item_storage[0]
                    item_typedef = item_typedef[0]
                typedef["identical"] = identical
                typedef["item_storage"] = item_storage
                typedef["items"] = item_typedef
            else:
                raise TypeError(typedef["type"])
        else: # Python object inside a Numpy structured dtype
            if isinstance(data, ndarray):
                storage, typedef2 = get_form_list(data)  #plain storage
            else:
                storage, typedef2 = get_form_python_inside_numpy(data)  #plain storage
            if isinstance(typedef2, dict):
                typedef.clear()
                typedef.update(typedef2)
            else:
                typedef = typedef2
    elif storage0 == "mixed-binary":
        assert isinstance(data, (ndarray, void)), type(data)
        assert isinstance(typedef, dict), type(typedef)
        if typedef["type"] == "object":
            return visit_typedef_numpy_struct(storage0, typedef, data)
        elif typedef["type"] == "array":
            raise Exception #should be impossible: arbitrary-length Numpy arrays inside Numpy dtypes are Python objects (storage0 = None)!
        elif typedef["type"] == "tuple":
            return visit_typedef_numpy_tuple(storage0, typedef, data)
        else:
            raise TypeError(typedef["type"])
    else:
        raise ValueError(storage0)
    return storage

def get_form_python_inside_numpy(data):
    if isinstance(data, list):
        return get_form_list_plain(data)
    elif isinstance(data, dict):
        return get_form_dict_plain(data)
    elif isinstance(data, _string_types):
        return "pure-plain", "string"
    elif data is None:
        return "pure-plain", "null"
    else:
        raise TypeError(type(data)) #must be list, dict, string or None; scalar Python objects not allowed

def get_tform_numpy_pyobject(dt):
    if dt.ndim:
        if dt.ndim > 1: raise NotImplementedError #TODO: multi-dimensional Python tuples
        storage = None #technically, mixed-binary, but this will be visited
        typedef = {
            "type": "tuple",
            "items": None, #to be visited
            "shape": dt.shape,
        }
    else:
        storage, typedef = None, None #to be visited
    return storage, typedef

def get_tform_numpy_builtin(dt):
    dtb = dt.base
    if dtb == object:
        return get_tform_numpy_pyobject(dt)

    storage = "pure-binary"
    if dtb == bool:
        typedef0 = "boolean"
    elif any([dtb == t for t in _integer_types]):
        typedef0 = "integer"
    elif any([dtb == t for t in _float_types]):
        typedef0 = "number"
    elif is_np_str(dtb):
        typedef0 = "string"
    else:
        raise TypeError(dtb)

    typedef = {
        "type": typedef0,
        "bytesize": dt.base.itemsize,
    }
    if typedef0 == "integer":
        unsigned = is_unsigned(dt)
        typedef["unsigned"] = unsigned
    if dt.ndim:
        typedef = {
            "type": "tuple",
            "items": typedef,
            "identical": True,
            "shape": dt.shape,
        }
    return storage, typedef

def get_tform_numpy_struct(dt):
    if not dt.isalignedstruct:
        raise TypeError("Composite dtypes must be memory-aligned", dt)
    if not dt.isnative:
        raise TypeError("Composite dtypes must be native")
    storages = {}
    props = {}
    typedef = {"type": "object", "properties": props, "bytesize": dt.itemsize}
    typedef["order"] = list(dt.fields)
    for fieldname in dt.fields:
        cstorage, ctypedef = get_tform_numpy(dt[fieldname])
        storages[fieldname] = cstorage
        props[fieldname] = ctypedef
    storage_set = set(storages.values())
    if len(storage_set) == 1 and storage_set.pop() == "pure-binary":
        storage = "pure-binary"
    else:
        storage = "mixed-binary"
        for fieldname in dt.fields:
            cstorage = storages[fieldname]
            if cstorage == "pure-binary":
                continue
            ctypedef = props[fieldname]
            if ctypedef is None or isinstance(ctypedef, str):
                ctypedef = {"type": ctypedef}
                props[fieldname] = ctypedef
            ctypedef["storage"] = cstorage
    if dt.ndim:
        typedef["storage"] = storage
        typedef = {
            "type": "tuple",
            "shape": dt.shape,
            "items":  typedef,
            "identical": True
        }
    return storage, typedef


def is_np_str(dt):
    return dt == np.dtype("S%d" % dt.itemsize) or dt == np.dtype("U%d" % (dt.itemsize/4))

def get_tform_numpy(dt):
    if dt.base.isbuiltin or is_np_str(dt) or dt in dt_builtins:
        return get_tform_numpy_builtin(dt)
    return get_tform_numpy_struct(dt)

def get_form_dict_plain(data):
    typedef = {"type": "object"}
    if not len(data):
        return "pure-plain", typedef
    props = {}
    typedef["properties"] = props
    storages = {}
    for k,v in data.items():
        cstorage, ctypedef = get_form(v)
        props[k] = ctypedef
        storages[k] = cstorage
    storage_set = set(storages.values())
    if len(storage_set) == 1 and storage_set.pop() == "pure-plain":
        storage = "pure-plain"
    else:
        storage = "mixed-plain"
        for fieldname in data:
            cstorage = storages[fieldname]
            if cstorage == "pure-plain":
                continue
            ctypedef = props[fieldname]
            if isinstance(ctypedef, str):
                ctypedef = {"type": ctypedef}
                props[fieldname] = ctypedef
            ctypedef["storage"] = cstorage
    return storage, typedef

def get_form_dict(data):
    if isinstance(data, Scalar):
        raise TypeError
    elif isinstance(data, void):
        dt = data.dtype
        storage, typedef = get_tform_numpy(dt)
        if storage in (None, "mixed-binary"):
            storage = visit_typedef_numpy(storage, typedef, data)
    elif isinstance(data, _array_types):
        raise TypeError
    elif isinstance(data, dict):
        storage, typedef = get_form_dict_plain(data)
    else:
        raise TypeError
    return storage, typedef

def get_form_items_numpy(data):
    dt = data.dtype
    storage, item_typedef = get_tform_numpy(dt)
    identical = True
    items = item_typedef
    if storage == "pure-binary":
        pass
    elif storage == "mixed-binary":
        items, identical = visit_typedef_numpy_items(item_typedef, data)
    elif storage is None: #pyobject
        return None, item_typedef, True #Will be visited later
    else:
        raise ValueError(storage)
    return storage, items, identical

def get_form_items_list_plain(data):
    identical = True
    storages = []
    items2 = []
    for d in data:
        item_storage, item_typedef = get_form(d)
        storages.append(item_storage)
        items2.append(item_typedef)

    storage_set = set(storages)
    if len(storage_set) > 1:
        different = True
    else:
        different = False
        for item_typedef in items2[1:]:
            if item_typedef != items2[0]:
                different = True
                break
    if different:
        items = items2
        identical = False
        if len(storage_set) == 1 and storage_set.pop() == "pure-plain":
            storage = "pure-plain"
        else:
            storage = "mixed-plain"
        for n in range(len(data)):
            s = storages[n]
            if s == "pure-plain":
                continue
            ctypedef = items[n]
            if isinstance(ctypedef, str):
                ctypedef = {"type": ctypedef}
                items[n] = ctypedef
            items[n]["storage"] = s
    else:
        if not len(items2):
            storage = "pure-plain"
            items = None
        else:
            items = items2[0]
            child_storage = storages[0]
            if child_storage == "pure-plain":
                storage = "pure-plain"
            else:
                storage = "mixed-plain"
                if isinstance(items, str):
                    items = {"type": items}
                items["storage"] = child_storage
    return storage, items, identical

def get_form_list(data):
    extra = {}
    if isinstance(data, ndarray):
        dt = data.dtype
        if not dt.isnative:
            raise TypeError("dtypes must be native")
        storage, items, identical = get_form_items_numpy(data)
        extra = {
            "shape": data.shape,
        }
        strides = data.strides
        contiguous = is_contiguous(data)
        extra["contiguous"] = contiguous
        if not contiguous:
            extra["strides"] = strides
    elif isinstance(data, _array_types):
        storage, items, identical = get_form_items_list_plain(data)
        extra = {
            "shape": (len(data),),
        }
    elif isinstance(data, dict):
        raise TypeError
    elif isinstance(data, Scalar):
        raise TypeError
    elif isinstance(data, void):
        raise TypeError
    else:
        raise TypeError

    typedef = {
        "type": "array",
        "items": items,
        "identical": identical
    }
    typedef.update(extra)
    return storage, typedef

def get_form_list_plain(data):
    #data must be a plain list
    storage, items, identical = get_form_items_list_plain(data)
    typedef = {
        "type": "array",
        "items": items,
        "identical": identical,
        "shape": (len(data),),
    }
    return storage, typedef


def get_form(data):
    if isinstance(data, bytes):
        data = np.array(data)
    if isinstance(data, Scalar):
        storage, typedef = "pure-plain", get_typedef_scalar(data)
    elif isinstance(data, void):
        dt = data.dtype
        storage, typedef = get_tform_numpy(dt)
        if storage in (None, "mixed-binary"):
            storage = visit_typedef_numpy(storage, typedef, data)
    elif isinstance(data, _array_types):
        storage, typedef = get_form_list(data)
    elif isinstance(data, dict):
        storage, typedef = get_form_dict_plain(data)
    else:
        raise TypeError(type(data))
    assert storage in (None, "pure-plain", "mixed-plain", "pure-binary", "mixed-binary"), storage
    return storage, typedef
