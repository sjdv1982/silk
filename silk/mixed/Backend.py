from copy import deepcopy
from collections import defaultdict

from .get_form import get_form as calc_form
from . import scalars

def get_subform(form, path):
    if not len(path):
        return form
    if form is None:
        return None
    attr = path[0]
    if isinstance(form, str):
       return None
    type_ = form["type"]
    if type_ == "object":
        if not isinstance(attr, str):
            raise AttributeError(attr)
        if "properties" not in form:
            return None
            #raise ValueError(form)
            #pass
        if attr not in form["properties"]:
            return None
        subform = form["properties"][attr]
    elif type_ in ("array", "tuple"):
        if not isinstance(attr, int):
            raise AttributeError(attr)
        if attr < 0:
            items = form["items"]
            attr = len(items) + attr
        if form["identical"]:
            subform = form["items"]
        else:
            items = form["items"]
            if attr >= len(items):
                return None
            subform = form["items"][attr]
    elif type_ in scalars:
        return None
    else:
        raise TypeError(type_)
    return get_subform(subform, path[1:])


class Backend:
    formless = False
    toplevel_dataclass = None
    def __init__(self, plain):
        self._plain = plain

    @property
    def plain(self):
        return self._plain

    def get_subform(self, path):
        for pp in path:
            if not isinstance(pp, (int, str)):
                raise AttributeError(path)
        form = self.get_form()
        return get_subform(form, path)

    def set_path(self, path, data):
        for pp in path:
            if not isinstance(pp, (int, str)):
                raise AttributeError(path)
        if data is None:
            return self.del_path(path)
        if self.formless:
           result = self._set_path(path, data)
           self._update(path)
           return result

        for start in range(len(path)-1):
            subpath = path[:start]
            subdata = self.get_path(subpath)
            if subdata is None:
                break
        else:
            start = len(path) - 1
        for n in range(start, len(path)-1):
            p = path[n]
            if isinstance(p, int):
                d = []
            else:
                d = {}
            self._set_path(path[:n], d)
        self._set_path(path, data)

        if not len(path) or start == len(path) - 1:
            newdata = data
            newpath = path
        else:
            newpath = path[:start+1]
            newdata = self.get_path(newpath)
        storage, form = calc_form(newdata)
        self._set_form(form, newpath)
        if self.plain:
            assert storage == "pure-plain", storage
        else:
            if not len(path):
                self._set_storage(storage)
        self._update(path)

    def del_path(self, path):
        for pp in path:
            if not isinstance(pp, (int, str)):
                raise AttributeError(path)
        if not len(path):
            self._del_path(())
            return
        subdata = self.get_path(path[:-1])
        if subdata is None:
            return
        subform, substorage = self._get_form(path[:-1])
        if isinstance(subform, str):
            subformtype = subform
        else:
            subformtype = subform["type"]
        attr = path[-1]
        if isinstance(attr, int):
            if subformtype != "array":
                raise TypeError(subform) #must be "array"
            if substorage.endswith("binary"):
                raise TypeError #cannot remove items from binary
            self._del_path(path)
        else:
            if subformtype != "object":
                raise TypeError(subform) #must be "object"
            self._del_path(path)
        self._update(path)

    def insert_path(self, path, data):
        for pp in path:
            if not isinstance(pp, (int, str)):
                raise TypeError(path)
        attr = path[-1]
        if not isinstance(attr, int):
            raise TypeError(path)
        subdata = self.get_path(path[:-1])
        if subdata is None:
            self.set_path(path[:-1], [])
        subform, storage = self._get_form(path[:-1])
        if isinstance(subform, str):
            subformtype = subform
            substorage = "pure-plain"
        else:
            subformtype = subform["type"]
            substorage = subform.get("storage", storage)
        if subformtype != "array":
            raise TypeError(subform) #must be "array"
        if substorage.endswith("binary"):
            raise TypeError #cannot insert items in binary
        self._insert_path(data, path)
        self._update(path)

class DefaultBackend(Backend):
    def __init__(self, plain, *, data_getter=None, data_setter=None):
        super().__init__(plain)
        self._data = None
        self._form = None
        self._storage = None
        self._data_getter = data_getter
        self._data_setter = data_setter

    def get_storage(self):
        if self._plain:
            return "pure-plain"
        assert isinstance(self._storage, (str, type(None)))
        return self._storage

    def _set_storage(self, value):
        if self._plain:
            raise AttributeError
        assert isinstance(value, (str, type(None)))
        self._storage = value

    def get_data(self):
        if self._data_getter is not None:
            return self._data_getter()
        else:
            return self._data

    def get_form(self):
        return self._form

    def get_path(self, path):
        for pp in path:
            if not isinstance(pp, (int, str)):
                raise TypeError(path)
        data = self.get_data()
        if not len(path):
            return data
        result = data
        for p in path:
            if result is None:
                return None
            if isinstance(p, int):
                if p >= len(result):
                    return None
            result = result[p]
        return result

    def _set_path(self, path, data):
        if not len(path):
            if self._data_setter is not None:
                self._data_setter(data)
            else:
                self._data = deepcopy(data)
            return
        subdata = self.get_path(path[:-1])
        attr = path[-1]
        if isinstance(attr, int):
            if attr >= len(subdata):
                if not isinstance(subdata, list):
                    raise TypeError(type(subdata))
                for n in range(len(subdata), attr + 1):
                    subdata.append(None)
        subdata[attr] = data

    def _insert_path(self, data, path):
        subdata = self.get_path(path[:-1])
        attr = path[-1]
        if not isinstance(attr, int):
            raise AttributeError(attr)
        if not isinstance(subdata, list):
            raise TypeError(subdata)
        if len(subdata) >= attr:
            for n in range(len(subdata), attr+1):
                subdata.append(None)
            subdata[attr] = data
        else:
            subdata.insert(attr, data)

    def _del_path(self, path):
        if not len(path):
            if self._data_setter is not None:
                self._data_setter(None)
            else:
                self._data = None
            return
        subdata = self.get_path(path[:-1])
        attr = path[-1]
        subdata.pop(attr)

    def _get_form(self, path):
        if not len(path):
            return self.get_form(), self.get_storage()
        subform, substorage = self._get_form(path[:-1])
        attr = path[-1]
        if isinstance(attr, int):
            if subform["identical"]:
                subform2 = subform["items"]
            else:
                subform2 = subform["items"][attr]
        else:
            subform2 = subform["properties"][attr]
        if "storage" in subform2:
            substorage = subform2["storage"]
        return subform2, substorage

    def _set_form(self, form, path):
        if not len(path):
            self._form = deepcopy(form)
            return
        subform, _ = self._get_form(path[:-1])
        attr = path[-1]
        if isinstance(attr, int):
            if subform["identical"]:
                subform["items"] = form
            else:
                subform["items"][attr] = form
        else:
            if "properties" not in subform:
                subform["properties"] = {}
            subform["properties"][attr] = form

    def _update(self, path):
        data = self.get_data()
        storage, form = calc_form(data)
        if self.plain:
            assert storage == "pure-plain", storage
        else:
            self._set_storage(storage)
            self._form = form

class SilkBackend(DefaultBackend):
    _silk = None
    def __init__(self):
        super().__init__(plain=False)

    def set_silk(self, silk):
        from ..Silk import Silk
        if not isinstance(silk, Silk):
            raise TypeError("silk")
        self._silk = silk

    def _update(self, path):
        super()._update(path)
        assert self._silk is not None
        data = self._silk
        for item in path:
            data = data._getitem(item)
        data.validate(full=None)

class StructuredCellBackend(Backend):
    def __init__(self, structured_cell):
        self._structured_cell = structured_cell
        super().__init__(False)
        self._calc_form()

    @property
    def formless(self):
        return self._structured_cell.hash_pattern is not None

    @property
    def toplevel_dataclass(self):
        sc = self._structured_cell
        if sc.hash_pattern is None:
            return None
        elif sc.hash_pattern == {"*": "#"}:
            return dict
        elif sc.hash_pattern == {"!": "#"}:
            return list
        else:
            raise NotImplementedError(sc.hash_pattern)


    def _calc_form(self):
        sc = self._structured_cell
        if sc.hash_pattern is not None:
            self._storage = None
            self._form = None
        else:
            data = self.get_data()
            storage, form = calc_form(data)
            if form == "null":
                form = None
            self._storage = storage
            self._form = form

    def get_storage(self):
        if self._plain:
            return "pure-plain"
        assert isinstance(self._storage, (str, type(None)))
        return self._storage

    def _set_storage(self, value):
        if self._plain:
            raise AttributeError
        assert isinstance(value, (str, type(None)))
        self._storage = value

    def get_data(self):
        return self._structured_cell._get_auth_path(())

    def get_form(self):
        return self._form

    def get_path(self, path):
        return self._structured_cell._get_auth_path(path)

    def _set_path(self, path, data):
        sc = self._structured_cell
        if not len(path):
            sc._set_auth_path((), data)
            return
        probe_sub = False
        if sc.hash_pattern is None:
            probe_sub = True
        else:
            if sc.hash_pattern == {"*": "#"}:
                if len(path) == 1:
                    if sc._auth_none():
                        sc._set_auth_path((), {})
                else:
                    probe_sub = True
            elif sc.hash_pattern == {"!": "#"}:
                if len(path) == 1:
                    if sc._auth_none():
                        sc._set_auth_path((), [])
                else:
                    probe_sub = True
            else:
                raise NotImplementedError(sc.hash_pattern)
        if probe_sub:
            subdata = self.get_path(path[:-1])
            attr = path[-1]
            subpath = path[:-1]
            if subdata is None:
                if isinstance(attr, int):
                    sc._set_auth_path(path[:-1], [])
                else:
                    sc._set_auth_path(path[:-1], {})
                subdata = self.get_path(path[:-1])
            if isinstance(attr, int):
                if attr >= len(subdata):
                    if not isinstance(subdata, list):
                        raise TypeError(type(subdata))
                    for n in range(len(subdata), attr):
                        sc._set_auth_path(subpath + (n,), None)
        sc._set_auth_path(path, data)

    def _insert_path(self, data, path):
        raise NotImplementedError # StructuredCell Silk wrapper does not support insertion

    def _del_path(self, path):
        sc = self._structured_cell
        sc._set_auth_path(path, None)

    def _get_form(self, path):
        if not len(path):
            return self.get_form(), self.get_storage()
        subform, substorage = self._get_form(path[:-1])
        if subform is None:
            return subform, substorage
        attr = path[-1]
        if isinstance(attr, int):
            if subform["identical"]:
                subform2 = subform["items"]
            else:
                subform2 = subform["items"][attr]
        else:
            subform2 = subform["properties"][attr]
        if "storage" in subform2:
            substorage = subform2["storage"]
        return subform2, substorage

    def _set_form(self, form, path):
        if not len(path):
            if form == "null":
                form = None
            self._form = deepcopy(form)
            return
        subform, _ = self._get_form(path[:-1])
        if subform is None:
            return
        attr = path[-1]
        if isinstance(attr, int):
            if subform["identical"]:
                subform["items"] = form
            else:
                subform["items"][attr] = form
        else:
            if "properties" not in subform:
                subform["properties"] = {}
            subform["properties"][attr] = form

    def _update(self, path):
        self._calc_form()
        sc = self._structured_cell
        sc._join_auth()

class StructuredCellSchemaBackend(StructuredCellBackend):
    formless = False
    def __init__(self, structured_cell):
        self._structured_cell = structured_cell
        Backend.__init__(self, True)
        self._storage = "pure-plain"
        data = self.get_data()
        if data is None:
            ### self.set_path((), {}) # .set_path can be async!
            data = {}
        storage, form = calc_form(data)
        assert storage == "pure-plain"
        self._form = form

    def get_storage(self):
        return "pure-plain"

    def _set_storage(self, value):
        raise AttributeError

    def get_data(self):
        return self._structured_cell._get_schema_path(())

    def get_path(self, path):
        return self._structured_cell._get_schema_path(path)

    def _set_path(self, path, data):
        sc = self._structured_cell
        if not len(path):
            sc._set_schema_path((), data)
            assert self.get_data() is data
            return
        subdata = self.get_path(path[:-1])
        attr = path[-1]
        subpath = path[:-1]
        if isinstance(attr, int):
            if attr >= len(subdata):
                if not isinstance(subdata, list):
                    raise TypeError(type(subdata))
                for n in range(len(subdata), attr):
                    sc._set_schema_path(subpath + (n,), None)
        sc._set_schema_path(path, data)

    def _insert_path(self, data, path):
        raise NotImplementedError # StructuredCell Silk wrapper does not support insertion

    def _del_path(self, path):
        sc = self._structured_cell
        sc._set_schema_path(path, None)

    def _update(self, path):
        sc = self._structured_cell
        data = self.get_data()
        _, form = calc_form(data)
        self._form = form
        sc._join_schema()
