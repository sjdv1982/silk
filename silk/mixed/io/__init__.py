from ..get_form import get_form
from .to_stream import to_stream
from .from_stream import from_stream
from .serialization import serialize, deserialize