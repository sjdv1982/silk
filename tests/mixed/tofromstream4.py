#string tests for mixed to/from file

import sys
import numpy as np
from silk.mixed.get_form import get_form
from silk.mixed.io import to_stream, from_stream

import json

arr = np.zeros(10)

dt = np.dtype([
    ("a", np.uint32),
    ("b", np.float32),
    ("str1", "S4"),
    ("str2", "S5"),
], align=True)
arr2 = np.zeros(2, dtype=dt)
data = {"arr":  arr, "arr2": arr2, "v": arr2[0], "z": 10}
storage, form = get_form(data)

stream = to_stream(data, storage, form)
newdata = from_stream(stream, storage, form)
print(newdata["arr"].dtype == data["arr"].dtype)
print(newdata["arr"] == data["arr"])
print(newdata["arr"] == data["arr"])
print(newdata["arr2"]["str1"].tobytes() == data["arr2"]["str1"].tobytes())
print(newdata["arr2"]["str2"].tobytes() == data["arr2"]["str2"].tobytes())
print(newdata["arr2"].tobytes() == data["arr2"].tobytes())
print()
data = b'somebuffer'
storage, form = get_form(data)
stream = to_stream(data, storage, form)
newdata = from_stream(stream, storage, form)
print(data, newdata)
print(data == newdata)
