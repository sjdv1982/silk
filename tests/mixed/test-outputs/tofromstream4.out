True
[ True  True  True  True  True  True  True  True  True  True]
[ True  True  True  True  True  True  True  True  True  True]
True
True
True

b'somebuffer' b'somebuffer'
True
